Allow delayed initialization of the BlockRestrictable.

permcody opened the pull request:
This PR allows us to delay initialization of BlockRestrictable until MeshModifiers are running. This allows a MeshModifier to make changes to the Mesh blocks (and potentially boundaries) that remain fixed once the Mesh has been prepared and BlockRestrictiable initializes. This solves the issue #7011 which doesn't have anything to do with `depend_on`.

This PR uncovered a more serious defect however in BlockRestrictable in that some members are being accessed before being initialized (not invalid memory, just incorrect or missing contents). I've added an assert to `BlockRestrictable::isBlockSubset()` that fails if uncommented. This needs to be fixed and is referenced in ticket #7088.

closes #7011

permcody committed:
Allow delayed initialization of the BlockRestrictable. This is necessary
if mesh modifiers depend on block boundary information changes.

closes #7011

aeslaughter commented on framework/src/base/BlockRestrictable.C:
```diff
@@ -27,6 +27,8 @@ InputParameters validParams<BlockRestrictable>()
   // Add the user-facing 'block' input parameter
   params.addParam<std::vector<SubdomainName> >("block", "The list of block ids (SubdomainID) that this object will be applied");
 
+  params.addPrivateParam<bool>("delay_initialization", false);
```
Do we want to do this via a parameter or should we do this with an additional constructor flag in BlockRestrictable. 

```
BlockRestrictable(..., bool initialize=True);
```

friedmud replied:
I like the additional constructor flag... this is something that should be decided by the designer of the System/Class that is inheriting from BlockRestrictable... and it's much more direct to pass it in the constructor.

permcody replied:
I'm not sure I agree for multiple reasons. First from a design prespective we tend to start seeing these default arguments piling up and then people ignore them because you need to pass parameters than a pile of default arguments which you kind of don't care about because.. hey, defaults!

More importantly though we don't always need it even for a particular class. It's not static. We use this particular MeshModifier in a lot of contexts and it works fine. What's special about the test case Andy provided is that he's specifically mod'ing the mod. Unless you are doing that you don't need to delay the initialization.

friedmud replied:
Ok - if all MeshModifiers aren't always delaying BoundaryRestrictable initialization... that's a reason to have it as a parameter...

This is off-topic but my issue with the private parameters is that there's absolutely ZERO documentation that they exist.  How is someone making a new class supposed to even know that they have the option of delaying BoundaryRestrictable?

permcody replied:
We could do better I suppose. Parameters aren't really documented either unless people write Doxygen around their signatures.

Continuing off-topic, I kind of like the idea of unifying our Interface headers as well. The reason is similar to that for unifying our object headers. If I were to change the constructor for this class I'd have to change it everywhere it's used. Right now, I can slip a new parameter in its validParams method and then start using it. Yes, applications are indeed making use of our Interfaces in some places so this problem is real. It's just not nearly as widespread.

friedmud replied:
Parameters _are_ documented.  They show up in `--dump`, Peacock and the syntax web-pages... EXCEPT for the private ones.

permcody replied:
Private ones are not meant to be set from the input file so they shouldn't show up in Peacock. We could do something different for the developer API though. I'm all for documenting these things better. Let's continue that discussion on a different issue or the documentation issue @aeslaughter is working on.

friedmud replied:
Yes: I don't want the private ones to show up there... I was just pointing out that there is currently no way for a developer to ever know that private parameters exist and that they can manipulate them.

I really think in this particular case that some documentation on the Wiki is warranted that explains what the option is and why you would use it.  But... well... we have lots of things that are undocumented :-)

aeslaughter replied:
First, we should add a documentation string to the private parameters. Then we can addthem to the yaml dump and do what we want with them, including not show them in Peacock. However, if they are in the yaml dump than we can show them somewhere (i.e., in a developer information section of the website).

friedmud commented on framework/include/meshmodifiers/MeshModifier.h:
```diff
@@ -56,11 +56,17 @@ class MeshModifier :
 
   std::vector<std::string> & getDependencies() { return _depends_on; }
 
+  bool forcePrepare() const { return _force_prepare; }
```
Should have named it `justGo()`  ;-)

friedmud commented on framework/src/meshmodifiers/MeshModifier.C:
```diff
@@ -19,6 +19,9 @@ InputParameters validParams<MeshModifier>()
 {
   InputParameters params = validParams<MooseObject>();
   params.addParam<std::vector<std::string> >("depends_on", "The MeshModifiers that this modifier relies upon (i.e. must execute before this one)");
+  params.addParam<bool>("force_prepare", false,
```
Does this need to be user-settable?  When will the user know to turn it on or off?

I think this should be a flag set in the constructor of the class for each MeshModifier.

permcody replied:
It does need to be user settable. You only need to force a prepare if you are running subsequent modifiers that need information from the current modifier (this is a rare case).

friedmud replied:
Is that documented somewhere?  This is fairly complicated... there needs to be at least a little bit of documentation on the Wiki explaining what this thing is...

permcody replied:
It's documented in the documentation string 😉 . Just keep reading...

friedmud commented on framework/src/meshmodifiers/SideSetsAroundSubdomain.C:
```diff
@@ -58,6 +62,9 @@ SideSetsAroundSubdomain::modify()
     mooseError("_mesh_ptr must be initialized before calling SideSetsAroundSubdomain::modify()!");
 
+  // Initialize the BlockRestrictable parent
+  initializeBlockRestrictable(_pars);
```
I don't like that the Class has to remember to do this itself.

Perhaps just make `MooseApp` do this before `modify()` is run?  If it's already initialized then it will just be skipped, right?

permcody replied:
That's not really better overall in terms of design since now an outside class has to initialize another unrelated class. It does however remove the burden from a future class designer from running into this issue. Right now this is the only MeshModifier inheriting from BlockRestrictable (believe it or not) so I didn't want to over design.

I think a better design might be to call a non-virtual on the BASE CLASS that initializes itself and then calls the existing "modify()" method. That keeps the initialization all internal.

friedmud replied:
I mean - you're already doing something similar with `forcePrepare()`... that could have been done inside the MeshModifier too...

I don't really see what you mean... we call `initialize()` and all kinds of other "initialization" routines on objects before asking them to compute all the time.

It's not over-designing... this design isn't right.  `modify()` should not be doing "initialization" stuff.

I will agree that it's a minor issue though (no design is perfect)...

permcody replied:
OK - then I'll go with my solution above. It seems like the BASE CLASS should be doing the following:

```
initialization of the boundary restrictable class if necessary
modify()
calling mesh->prepare() if force
```

friedmud replied:
Sounds reasonable

permcody committed:
Refactor MeshModifiers a bit

This adds a new BASE CLASS method that handles initialization, modification
and forced preparation as needed instead of having all of that logic coded
in the MooseApp class.

refs #7011

permcody commented:
All comments should be addressed except for this ongoing private param issue.

aeslaughter commented on framework/src/base/BlockRestrictable.C:
```diff
@@ -201,6 +221,8 @@ BlockRestrictable::hasBlocks(std::vector<SubdomainID> ids) const
 bool
 BlockRestrictable::hasBlocks(std::set<SubdomainID> ids) const
 {
+  mooseAssert(_initialized, "BlockRestrictable not initialized");
```
Should we add a test that hits one of these asserts?

permcody replied:
Nah, There's little point in doing that. You'd have to manufacture a broken
object. There's nothing wrong with not testing asserts. They are there
because we expect them to never be hit, but they are good indicators when
things do go wrong and your assumptions break down.

On Thu, May 26, 2016 at 2:27 PM Andrew E Slaughter <notifications@github.com>
wrote:

> In framework/src/base/BlockRestrictable.C
> <https://github.com/idaholab/moose/pull/7089#discussion_r64815027>:
>
> > @@ -201,6 +221,8 @@ BlockRestrictable::hasBlocks(std::vector<SubdomainID> ids) const
> >  bool
> >  BlockRestrictable::hasBlocks(std::set<SubdomainID> ids) const
> >  {
> > +  mooseAssert(_initialized, "BlockRestrictable not initialized");
>
> Should we add a test that hits one of these asserts?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/7089/files/d99e945961c595f44e89f98e15de5d0c2c23ef19#r64815027>
>
